main/src/library/List.flix

```diff
@@ -0,0 +1,612 @@
+    /// If patching occurs at index `i+j` in `ys`, then the element at index `j` in `xs` is used.
+    ///
+    def patch[a](i: Int32, n: Int32, xs: List[a], ys: List[a]): List[a] = patchHelper(i, n, drop(-i, xs), ys, 0)
+
+    ///
+    /// Helper function for `patch`.
+    ///
+    def patchHelper[a](i: Int32, n: Int32, xs: List[a], ys: List[a], c: Int32): List[a] = match (xs, ys) with {
+        case (_, Nil) => ys
+        case (Nil, _) => ys
+        case (x :: qs, y :: rs) => if (c >= i && c < i+n) x :: patchHelper(i, n, qs, rs, c+1) else y :: patchHelper(i, n, xs, rs, c+1)
+    }
+
+    ///
+    /// Returns all permutations of `xs` in lexicographical order by element indices in `xs`.
+    /// That is, `xs` is the first permutation and `reverse(xs)` is the last permutation.
+    ///
+    def permutations[a](xs: List[a]): List[List[a]] = match xs with {
+        case Nil => Nil :: Nil
+        case _ => permutationHelper(0, xs)
+    }
+
+    ///
+    /// Helper function for `permutations`.
+    /// Returns all permutations of `xs` starting with an element at or after index `i`.
+    ///
+    def permutationHelper[a](i: Int32, xs: List[a]): List[List[a]] =
+        if (i == length(xs)) Nil
+        else append(applyHelper(at(i, xs), permutations(removeIndex(i, xs))), permutationHelper(i+1, xs))
+
+    ///
+    /// Returns all subsequences of `xs` in lexicographical order by element indices in `xs`.
+    /// That is, `xs` is the first subsequence and `Nil` is the last subsequence.
+    ///
+    def subsequences[a](xs: List[a]): List[List[a]] = match xs with {
+        case Nil => Nil :: Nil
+        case x :: rs => append(applyHelper(x, subsequences(rs)), subsequences(rs))
+    }
+
+    ///
+    /// Helper function for `permutations` and `subsequences`.
+    /// Returns `xs` with `x` added to the beginning of each element of `xs`.
+    ///
+    def applyHelper[a](x: a, ps: List[List[a]]): List[List[a]] = match ps with {
+        case Nil => Nil
+        case p :: rs => (x :: p) :: applyHelper(x, rs)
+    }
+
+    ///
+    /// Returns `xs` with `x` inserted between every two adjacent elements.
+    ///
+    def intersperse[a](x: a, xs: List[a]): List[a] = match xs with {
+        case x1 :: x2 :: rs => x1 :: x :: intersperse(x, x2 :: rs)
+        case _ => xs
+    }
+
+    ///
+    /// Returns the concatenation of the elements in `ys` with the elements of `xs` inserted between every two adjacent elements.
+    /// That is, returns y1 + x + y2 + x + ... + x + yn.
+    ///
+    def intercalate[a](xs: List[a], ys: List[List[a]]): List[a] = match ys with {
+        case Nil => Nil
+        case y :: Nil => y
+        case y1 :: y2 :: rs => append(append(y1, xs), intercalate(xs, y2 :: rs))
+    }
+
+    ///
+    /// Returns the transpose of `xs`.
+    /// Returns `xs` if the dimensions of the elements of `xs` are mismatched.
+    ///
+    def transpose[a](xs: List[List[a]]): List[List[a]] = if (!uniformHelper(xs)) xs else match xs with {
```

Comment:
As of now, in the case that the dimensions of `xs` are mismatched, `xs` is returned.  Should some sort of error occur instead?